elements/map/src/tileGrid.ts

```diff
@@ -0,0 +1,45 @@
+import { createXYZ } from "ol/tilegrid";
+import WMTSTileGrid from "ol/tilegrid/WMTS";
+import { getTopLeft, getWidth } from "ol/extent.js";
+import { get as getProjection } from "ol/proj.js";
+
+/**
+ * generates a WMTS tile grid for WMTS layers or else an XYZ tile grid, if defined.
+ * @param layer
+ * @returns
+ */
+export function generateTileGrid(layer: import("./generate").EoxLayer) {
+  let tileGrid;
+  if (!layer.source?.tileGrid) {
+    return undefined;
+  }
+
+  if (layer.source.tileGrid) {
+    if (layer.source.type === "WMTS") {
+      const projection = getProjection("EPSG:3857");
+      const projectionExtent = projection.getExtent();
+      const size = getWidth(projectionExtent) / 128;
+      const resolutions = new Array(19);
+      const matrixIds = new Array(19);
+      for (let z = 0; z < 19; ++z) {
+        // generate resolutions and matrixIds arrays for this WMTS
+        resolutions[z] = size / Math.pow(2, z);
+        matrixIds[z] = z;
+      }
+
+      tileGrid = new WMTSTileGrid({
+        ...layer.source.tileGrid,
```

Comment:
After discussing with Lubo, we think having the spread of the tileGrid at the end (after matrixIds) would allow overwriting some of the generated attributes, for very special strange cases that would be beneficial, could you please move it if you don't see any issue with it?